howto-dapp/src/main/java/de/timolia/howto/teamler/Teamler.kt

```diff
@@ -0,0 +1,127 @@
+            val rank = rankHistory.values.toTypedArray()[i]
+            Validate.notNull(rank, "Der " + (i + 1) + ". Rang von '" + name + "' existiert nicht")
+        }
+        val dates = rankHistory.keys
+                .map { s -> s.replace("hidden-", "") }
+                .map { date -> TeamlerRankChange.toDate(date) }
+                .sorted()
+        for (i in Math.min(1, dates.size - 1) until dates.size) {
+            var hidden = false
+            var rankOldKey: String = TeamlerRankChange.toString(dates[Math.max(i - 1, 0)])
+            if (!rankHistory.containsKey(rankOldKey) && rankHistory.containsKey("hidden-$rankOldKey")) {
+                rankOldKey = "hidden-$rankOldKey"
+            }
+            val rankOld = rankHistory[rankOldKey]
+            var rankNewKey: String = TeamlerRankChange.toString(dates[i])
+            if (!rankHistory.containsKey(rankNewKey) && rankHistory.containsKey("hidden-$rankNewKey")) {
+                rankNewKey = "hidden-$rankNewKey"
+                hidden = true
+            }
+            val rankNew = rankHistory[rankNewKey]
+            if (includeHidden || !hidden && rankNewKey != "initial") {
+                val teamlerRankChange = TeamlerRankChange(name, uuid, rankOld!!, rankNew!!, TeamlerRankChange.toString(dates[i]), hidden)
+                rankChanges.add(teamlerRankChange)
+            }
+        }
+        return rankChanges
+    }
+
+    fun getRankCurrent(): Rank {
+        if (rankCurrent == null) {
+            val teamlerRankChanges = getRankChanges(true)
+            val lastRankChange = teamlerRankChanges[teamlerRankChanges.size - 1]
+            rankCurrent = lastRankChange.rankTo
+        }
+        return rankCurrent!!
+    }
+
+    fun compare(o: Teamler): Int {
+        if (getRankCurrent() != o.getRankCurrent()) {
+            // compare by rank
+            return getRankCurrent().compare(o.getRankCurrent())
+        }
+
+        // compare by name
+        val rules = "< '_'"
+        return try {
+            val ruleBasedCollator = RuleBasedCollator(rules)
+            ruleBasedCollator.compare(name.lowercase(Locale.getDefault()), o.name.lowercase(Locale.getDefault()))
+        } catch (e: ParseException) {
+            e.printStackTrace()
+            throw RuntimeException()
+        }
+    }
+
+    fun hasResponsibilityMain(responsibility: String?): Boolean {
+        return hasResponsibility(responsibility, Stream.concat(if (responsibilitiesMain == null) Stream.empty() else responsibilitiesMain!!.stream(), responsibilitiesMainHidden?.stream()
+                ?: Stream.empty()).collect(Collectors.toList()))
+    }
+
+    fun hasResponsibilitySecondary(responsibility: String?): Boolean {
+        return hasResponsibility(responsibility, Stream.concat(if (responsibilitiesSecondary == null) Stream.empty() else responsibilitiesSecondary!!.stream(), responsibilitiesSecondaryHidden?.stream()
+                ?: Stream.empty()).collect(Collectors.toList()))
```

Comment:
(question/suggestion) not quite sure whats happning here but i tried to de-obfuscate it a little bit
```suggestion
        return hasResponsibility(
            responsibility,
            (responsibilitiesSecondary ?: emptyList()) + (responsibilitiesSecondaryHidden ?: emptyList())
        )
```